### PR TITLE
SNOW-940615: fix pyarrow import if pyarrow iterator is not compiled

### PR DESCRIPTION
**Pull request by sfc-gh-aling**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes #NNNN

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am modifying authorization mechanisms
   - [ ] I am adding new credentials
   - [ ] I am modifying OCSP code
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   Please write a short description of how your code change solves the related issue.

### CODE REVIEW

File: src/snowflake/connector/result_batch.py [hunk 1]
```diff
@@ -65,7 +62,9 @@ def _create_vendored_arrow_iterator(
     numpy: bool,
     number_to_decimal: bool,
     row_unit: IterUnit,
-) -> PyArrowIterator:
+):
+    from .arrow_iterator import PyArrowIterator
```
Comment (sfc-gh-yixie) [inline src/snowflake/connector/result_batch.py:66]: Will this raise an error if PyArrow isn't compiled?

File: tox.ini [hunk 2]
```diff
@@ -96,6 +96,14 @@ passenv = {[testenv]passenv}
 commands =
     {env:SNOWFLAKE_PYTEST_CMD} -m "not skipolddriver" -vvv {posargs:} test
 
+[testenv:noarrowextension]
+basepython = python3.8
```
Comment (sfc-gh-mkeller) [inline tox.ini:100]: Without this `tox` will install the library anyways
```suggestion
basepython = python3.8
skip_install = True
```

File: setup.py [hunk 3]
```diff
@@ -42,6 +42,10 @@
 extensions = None
 cmd_class = {}
 
+DISABLE_COMPILE_ARROW_EXTENSIONS = os.environ.get(
+    "DISABLE_COMPILE_ARROW_EXTENSIONS", False
+)
```
Comment (sfc-gh-mkeller) [inline setup.py:47]: Be careful, your code returns a `bool | str`. why not be explicit? Also follow the new env variable prefix
```suggestion
SNOWFLAKE_DISABLE_COMPILE_ARROW_EXTENSIONS = os.environ.get(
    "SNOWFLAKE_DISABLE_COMPILE_ARROW_EXTENSIONS", "false"
).lower() in ("y", "yes", "t", "true", "1", "on")
```

File: setup.py [hunk 4]
```diff
@@ -56,7 +60,7 @@
     )
     _ABLE_TO_COMPILE_EXTENSIONS = False
 
-if _ABLE_TO_COMPILE_EXTENSIONS:
+if _ABLE_TO_COMPILE_EXTENSIONS and not DISABLE_COMPILE_ARROW_EXTENSIONS:
```
Comment (sfc-gh-mkeller) [inline setup.py:63]: If you accept my other suggestion
```suggestion
if _ABLE_TO_COMPILE_EXTENSIONS and not SNOWFLAKE_DISABLE_COMPILE_ARROW_EXTENSIONS:
```

File: src/snowflake/connector/result_batch.py (hunk 1 above)
Comment (sfc-gh-aling) [reply src/snowflake/connector/result_batch.py:66]: no, the `_create_vendored_arrow_iterator` and `_create_nanoarrow_iterator` will only be called when we detect that arrow can be used.

**Review by sfc-gh-mkeller — CHANGES_REQUESTED** [review]

**Review by sfc-gh-mkeller — APPROVED** [review]
:ship: :ship:

**Pull request merged**